### DISCUSSION TITLE
[NCR#1 Proposal] Trustless Open-Source Remuneration through zkEmail

### DISCUSSION THREAD

**Discussion by wakeuplabs** [noir-lang/noir#5819] [NRG#1 zkEmail]
## **Summary**

Open-source contributors often face challenges in securing guaranteed monetary rewards even after resolving issues that have been pre-approved by organizations.

As observed in traditional bug bounty platforms like Code4Arena and open-source projects, developers may propose fixes or new versions only to be denied compensation despite clear issue severity definitions.

With zkEmail technology, we can address this **TRUST** issue. By standardizing how organizations create issues in repositories and linking the resolution process to zkEmail proofs, contributors can trustlessly receive payments once their proposed solutions are merged. Our approach ensures that when a contribution is accepted, a zk proof generated from an email (involving GitHub and the organization) triggers the release of funds from an escrow smart contract. This solution enables fair and private remuneration for open-source contributions.

## **Methodology**

Our first step is to validate that the zkEmail proof can be implemented in Noir. Once confirmed, we will incrementally add features, prioritizing what can realistically be achieved within the timeframe allocated by our research team. The process will emphasize clear documentation, benchmark comparisons, performance analysis, and a detailed step-by-step guide outlining the implementation.

The POC aims to enable an open-source project owner to create an issue that a contributor resolves. Once the project owner successfully merges the solution, a zk proof generated from the confirmation email will be used to trigger the release of funds from an escrow smart contract. If development progresses smoothly, we will explore how this technology can be applied directly to Aztec projects or within the broader zkEmail ecosystem. Despite Aztec’s ongoing development, we will also evaluate how this solution could integrate with the Aztec Blockchain.

The project team consists of two developers working closely for two months, alongside a project manager with strong technical research skills who will handle communication and progress tracking.

All team members have extensive knowledge in ZK and Blockchain, ensuring a well-executed and technically sound delivery.

![Photo from Telegram (2)](https://github.com/user-attachments/assets/b339ddc9-8b4b-45e3-82c6-1310c1dad6ad)

### Timeline and Deliverables

To accommodate evolving requirements and unforeseen challenges, we propose an iterative approach grounded in agile methodologies, particularly Scrum. The project will be led by a team of two developers and one technical lead, who will define the scope and deliverables for each iteration. The focus will be on delivering incremental value through well-documented outputs, regular demos, and progress updates shared with the Noir community. By conducting bi-weekly Scrum sprints, this strategy ensures ongoing adaptability, transparency, and alignment with the project’s objectives.

### First Month

**Phase 1: Planning and Setup (1 week)**

- **Deliverables:**
    - A comprehensive project plan with clearly defined milestones.
    - Configuration and setup of the development environment.
    - Finalized technical specifications and project requirements.

**Phase 2: Porting and Development (2 weeks)**

- **Deliverables:**
    - Migration of [Circom circuits](https://github.com/zkemail/email-wallet/tree/main/packages/circuits) to Noir.
    - Initial proof of concept (POC) for the core application.

### Second Month

**Phase 3: Feature Enhancement and Iteration (2 weeks)**

- **Deliverables:**
    - Iterative improvements to the POC, focusing on Noir library requirements.
    - Continuous testing environment setup for ZK chains, such as Aztec or alternative platforms.

**Phase 4: Integration and Testing (2 weeks)**

- **Deliverables:**
    - User interface for module creation and management.
    - Implementation of the Relayer service.
    - User interface for monitoring module operations.

**Phase 5: Deployment and Review (1 weeks)**

- **Deliverables:**
    - Deployment of all system components.
    - Completion of final documentation, including user guides.
    - Comprehensive project review with performance benchmarks and recommendations for future enhancements.

## **Team**

**WakeUp Labs** is a software development studio offering advanced development services to EVM-Compatible Chains, DAOs, and businesses. Our clients include Optimism, Arbitrum, Num Finance, 0x Protocol, Win Investments, and more.

We provide comprehensive development services from concept and strategy through to design and implementation, enabling our clients to unlock the full potential of technology.

With a proven track record of launching over 45 projects across Latin America, the USA, and Europe, WakeUp Labs is your trusted partner for innovation and growth.

- **Website:** [https://wakeuplabs.io](https://wakeuplabs.io)
- **GitHub:** [github.com/wakeuplabs](https://github.com/wakeuplabs)
- **Email:** [gonza@wakeuplabs.io](mailto:gonza@wakeuplabs.io)

### **Team Members**

***Milton Berman***, co-founder and CEO, brings extensive experience as a CTO in a decentralized identity project supported by IDB Lab and Bitcoin NGO. He also served as Product Owner at Rootstock Blockchain and as CTO of a mobile game studio. (GitHub: [WakeUp](https://github.com/wakeuplabs))

***Elías Molini*** is a seasoned full-stack developer with deep expertise in smart contracts and ZK technologies. He has received multiple grants, is an active contributor to open-source projects, and serves as a ZK expert at WakeUp Labs. (GitHub: [elmol](https://github.com/elmol))

***Christian Escalante*** has over 5 years of experience in blockchain development. He is the former Technical Lead at RSK Blockchain and the current CTO of WakeUp Labs. (GitHub: [chescalante](https://github.com/chescalante))

## **Start Date**

- 9 September 2024

**Comment by Savio-Sou** [comment]
Selected proposals for NCR#1 are announced: https://github.com/orgs/noir-lang/discussions/5932

Thank you for your support on the initiative. NCR is just one of the many ways to start your explorations with Noir, check the announcement for more potential ways to collaborate.

Personally resonates a lot with the project, do let me know if the team is exploring it.

One way or another, let's bring the world ZK!
